### ISSUE TITLE
Discussion: Future of prune and rebuild-index

### ISSUE THREAD

**Issue by aawsome** [restic/restic#2547]
I would like to discuss how to improve the actual situation of `prune` (and `rebuild-index` which is used within `prune`) as there are quite some issues about it.

So far I see the following issues:

1. It uses too much memory (#1723). This is because of general memory problems with the index in `internal/repository`, see #1988 but also because `internal/index` is used additional to the first index structure. This is the reason most users start complaining about memory consumption when using `prune`.
2. It is too slow, see  #1599, #2024, #2162. This is mainly because it reads all pack files (twice?). This completely makes 'prune' unusable for remote repositories with low bandwidth. Also not parallelized (#1470).
3. It is not customizable. Repacking is slow/expensive but cannot be turned of or customized (#1985, #2305)
4. It does too many things and is too complicated (#2227). It first rebuilds the index from pack files, then walks all snapshots and all dirs within to get the used blobs, then finds the packs where to delete these blobs and rewrites all those packs. Finally the index is again re-created completely from the resulting pack files. Hence it not only removes things, but reorganizes the repository and is even able to recover from some broken repo issues. However this is one source for the performance problems.

There are already some proposals to fix specific issues, see #1994, #2340, #2507.

I also started a rebuild of prune functionalities in #2513. Here I only used the index from `internal/repository` to clean unused blobs from the index, remove unused blobs and optionally repack if requested by the user. This solves IMO Points 1. to 3. and has already been successfully used by me and others for production repostitories.

About 1.: Using #2513 and the new commands, `internal/index` is no longer used and could also be removed if `prune` is substituted by these commands.
About 4.: There is so far a command missing to "repack" the index files if wanted (to get rid of small index files) and there is no way implemented to recover the index from the pack files. (i.e. the functionality of `rebuild-index`)
I can propose to work on these issues in #2513, if wanted.

So my question is: Where should `prune` and `rebuild-index` go to?
Is it a way to complete #2513 and then use all of this functionality in a new `prune` command?
It would be great to get some direction here!

**Comment by pmkane** [comment]
Interested in this as well.

Sadly we had issues with #2340 and repo corruption, so had to discontinue our testing.

**Comment by eigood** [comment]
Perhaps one should treat pruning like the java GC process?  Mark/Sweep?  Or any other technology terms from that problem space?

**Comment by aawsome** [comment]
@pmkane You can use #2513 and run `restic cleanup-index` followed by `restic cleanup-packs`. Without repacking (i.e. without the `--unused-percent`, `--unused-space`, `--used-space` flags) this has been tested quite often and I use it for my production repos.

@eigood I agree that this should be discussed. It mainly goes in the direction of having lock-free pruning. Please note that on repository-level there is no possibility to modify files as the file name equals the SHA256 hash of the content. Hence every "change" must be done by creating new files and deleting the old ones. Your proposal about mark/sweep hence mainly regards the final deletion of old files. This can IMO be implemented independently from the improvements I'm trying to solve within this issue.

About the "repacking" of index files: This can not easily be done with the actual index implementation in `internal/repository`. I added a command `repack-index` to #2513 using only the index implementation of `internal/index`.

So in my opinion #2513 is now usable to replace `prune` funcionality. However, the used index implementations are not optimal. This is an issue I think should be solved after the work in #2523.

So the next steps are:
 
- finish reimplementation of index (#2523)
- use new index implementation in `cleanup-index`, `cleanup-packs` and `repack-index`
- combine the three commands to a single new `prune` command
- (optional) remove the three commands

In meanwhile #2513 can be used for standard pruning operation.

**Comment by pmkane** [comment]
We'll give it a whirl.

**Comment by pmkane** [comment]
Works well in our testing and merges cleanly against #2195, which allows a single restic binary that can:

1) Backup at wire speed
2) Restore at wire speed
3) Prune large repos quickly

Would be great to see prune switched to use this code and to have both this and #2195 merged.

**Comment by MichaelEischer** [comment]
@aawsome Thanks for your work on a faster and less memory-hungry prune implementation.

My take on `rebuild-index` is that it should be primarily used to repair a corrupted repository index. An error-resilient version of `prune` should be capable of handling most other cases.

Beside the referenced issues there's also been some discussion on a lock-less two phase compaction in #1141. My understanding of that issue is that it's mostly orthogonal to this issue as it primarily affects the selection of which blobs may be pruned.

Merging the three individual commands into a single one is a _must_ in my opinion. In order to make the individual commands safe for an enduser, each of them would have to run a full set of sanity checks in the beginning. In addition the separation currently means rewriting large parts of the index three times in a row. A single command could handle that much better and just skip a few operations if told so by the user.

With `repack-index` I get the impression that it mostly solves a problem caused by the implementation of `cleanup-index` and `cleanup-packs`. Especially as the improved index implementation from #2523 is rather oblivious to a large number of small index files. Thus it would be sufficient to just merge indexes which have to be modified anyways.

The implementation in #2513 seems to be only partially able to handle damaged repositories. This is a problem that would need to be solved, even if that would mean to just refuse doing anything in case of repository damages similar to the current prune implementation.

In my opinion the handling of potential damages should look like the following:
In particular a prune implementation should only delete packs if it can _prove_ that these are no longer necessary. I know of two ways to do this: Either restic has found all used blobs in the repository and can therefore tell that it just needs to keep certain packs and can thus delete everything else (this also includes damaged or otherwise invalid packs). The other option would be that restic checks that a pack is unused by listing the content of the pack (this is also possible by looking at the index, if it contains a complete list of the blobs in a pack) and only removing it if no blob in that pack is used any more. The first option, which looks at the whole repository, is implemented in the current prune command, but that makes it impossible to prune damaged repositories. The second option would allow to do that, but it wouldn't be able to remove packs with unknown content (either damaged or otherwise invalid). The latter files could only be removed by combining this with option one. The reasoning here is that is might not be possible to read an otherwise valid pack due to temporary backend problems and therefore restic shouldn't blindly delete such pack files.

In case a blob exists in multiple packs, restic must make sure to retain a valid copy of the blob and explicitly verify its integrity. Blobs from packs that are to be rewritten must also be verified. The intuition here is to verify all blobs that are touched in any way by the prune operation. All other parts of the repository are not affected by the prune operation and therefore there's no need to check these. 

I'd also like to add a fifth and sixth problem of the current prune implementation:

-  It needs huge amounts of free storage space to complete and is not able to continue interrupted prune runs. A prune run only deletes data in the last step to avoid data loss if the prune run is interrupted. (#2473 #2343 #2112)
- It is totally stuck when there's no free disk space (#1140)

This problem could be solved by implementing prune to work incrementally: Basically it could process all packs referenced by one index file and then replace the index and remove no longer necessary packs. Advanced versions could also merge indexes on-the-fly and try to do something smart with packs that are referenced by multiple indexes.

The disk space issue probably arises most of the time due to a backup run (assuming that prune works incrementally). Such a situation could be resolved by just deleting full unused indexes and their packs as a first pass.

**Comment by aawsome** [comment]
@MichaelEischer  Thank you for your comments. I tried to make a new PR, #2718 which combines the three commands. There are still some open issues. However I would be happy to get comments as soon as possible during development, so feel free to have a look at the new implementation.

**Comment by whereisaaron** [comment]
Fantastic thread. The `prune` performance is a sore point for us too. It is incredibly vulnerable to latency. And even while this major work is looked at, smaller changes could help a lot. E.g. sitting and watching Restic deleting packs, one at a time, at a rate of about one pack per second, is viscerally painful :-)

Restic appears to know all the packs that need to be deleted, Could more that one pack be deleted in parallel? Linear speed-up seem possible here, deleting two packs at one would halve the time for this step, from 8-9 hours to 4 hours. Just that is large gain.
```
[20:10] 3.85%  1164 / 30227 packs deleted
```

**Comment by aawsome** [comment]
@whereisaaron you might want to try #2718 (please on a copy of your repo!). The actual state should speed up pure deletion by a factor of 8. But IMO the other improvements will speed up your pruning even *much more massively*!

**Comment by aawsome** [comment]
I think that #2718 is almost finished now, hence I closed #2513. 

I will now change #2718 such that it replaces the present command `prune`, work on tests and documentation and try to find a reviewer :wink:

**Comment by aawsome** [comment]
Just announcing that I changed the state of #2718 to "ready for review".
And first performance tests look very promising... if you are willing to do your own tests, I'm very curious about your results!

**Comment by zsau** [comment]
Nitpick: "cacheable" is a much more common spelling than "cachable" in my experience (and Google seems to agree).

**Comment by aawsome** [comment]
@zsau Thanks for the hint - is changed now.
